File: helper/darwin/KeeWebHelper/KeyRunner.m
#import "KeyRunner.h"

@implementation KeyRunner

CGEventSourceRef eventSource = nil;

+ (void)keyPress:(unichar)ch code:(CGKeyCode)code flags:(CGEventFlags)flags {
    [self keyUpDown:ch code:code flags:flags down:false];
    [self keyUpDown:ch code:code flags:flags down:true];
}

+ (void)keyUpDown:(unichar)ch code:(CGKeyCode)code flags:(CGEventFlags)flags down:(bool)down {
    if (!eventSource) {
        eventSource = CGEventSourceCreate(kCGEventSourceStateHIDSystemState);
    }
    CGEventRef keyEvent = CGEventCreateKeyboardEvent(eventSource, code, down);
    if (ch) {
        CGEventKeyboardSetUnicodeString(keyEvent, 1, &ch);
    }
    if (flags) {
        CGEventSetFlags(keyEvent, flags);
    }
    CGEventPost(kCGHIDEventTap, keyEvent);
    CFRelease(keyEvent);
}

@end

File: helper/darwin/KeeWebHelper/KeeWebHelper/KeyRunner.swift
import Foundation

class KeyRunner {
    static let inputSource = CGEventSourceCreate(CGEventSourceStateID.HIDSystemState)

    class func keyPress(char: UniChar?,
                        code: CGKeyCode = 0,
                        flags: CGEventFlags? = nil) {
        keyUpDown(char, code: code, flags: flags, down: true)
        keyUpDown(char, code: code, flags: flags, down: false)
    }
    
    class func keyUpDown(char: UniChar?,
                         code: CGKeyCode,
                         flags: CGEventFlags?,
                         down: Bool) {
        let keyEvent = CGEventCreateKeyboardEvent(KeyRunner.inputSource, code, down)
        if (char != nil) {
            var ch = char!;
            CGEventKeyboardSetUnicodeString(keyEvent, 1, &ch)
        }
        if (flags != nil) {
            CGEventSetFlags(keyEvent, flags!)
        }
        CGEventPost(CGEventTapLocation.CGHIDEventTap, keyEvent)
    }
}
